File: Data Structures & Algorithms/implement-stack-using-queues/submission-0.go
type MyStack struct {
	head *ListNode
}

type ListNode struct {
	val int
	next *ListNode
}

func Constructor() MyStack {
	return MyStack{}
}

func (this *MyStack) Push(x int) {
	n := ListNode{
		val: x,
	}
	if this.head != nil {
		n.next = this.head
	}
	this.head = &n
}

func (this *MyStack) Pop() int {
	v := this.head.val
	if this.head.next != nil {
		this.head = this.head.next
	}
	return v
}

func (this *MyStack) Top() int {
	return this.head.val
}

func (this *MyStack) Empty() bool {
	if this.head == nil {
		return true
	}
	return false
}


/**
 * Your MyStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param2 := obj.Pop();
 * param3 := obj.Top();
 * param4 := obj.Empty();
 */

File: Data Structures & Algorithms/design-browser-history/submission-0.py
class BrowserHistory:

    class HistoryEntry:
        url: str
        next: HistoryEntry
        prev: HistoryEntry

        def __init__(self, url: str):
            self.url = url
            self.next = None
            self.prev = None


    def __init__(self, homepage: str):
        self.curr = self.HistoryEntry(homepage)

    def visit(self, url: str) -> None:
        newHe = self.HistoryEntry(url)
        newHe.prev = self.curr
        self.curr.next = newHe
        self.curr = newHe

    def back(self, steps: int) -> str:
        for i in range(steps):
            if self.curr.prev:
                self.curr = self.curr.prev
            else:
                break
        return self.curr.url

    def forward(self, steps: int) -> str:
        for i in range(steps):
            if self.curr.next:
                self.curr = self.curr.next
            else:
                break
        return self.curr.url
        


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)